docs_src/design/adr/014-Secret-Provider-For-All.md

```diff
@@ -0,0 +1,398 @@
+
+The above example results in the Redis credentials being copied into app-service-xyz's `SecretStore` at `/secret/edgex/app-service-xyz/redis`.
+
+Similar approach could be taken for Message Bus credentials where a common `SecretStore` is created with the Message Bus credentials saved. The services request the credentials are copied into their exclusive `SecretStore` using `common/messagebus` as the secret identifier.
+
+Full specification for the environment variable's value is a comma separated list of service entries defined as:
+
+```
+<service-name>[optional list of known secret IDs sperated by ;]>
+```
+
+Example with one service specifying IDs for know secrets and one without known secrets 
+
+```yaml
+ADD_SECRETSTORE: "appservice-xyz[appservice/redisdb; common/messagebus], appservice-http-export"
+```
+
+When the `ADD_SECRETSTORE` environment variable is processed to create these `SecretStores`, it will copy the specified saved secrets from the initial `SecretStore` into the service's `SecretStore`. This all depends on the completion of database or other credential bootstrapping and the secrets having been stored prior to the environment variable being processed. security-secretstore-setup will need to be refactored to ensure this sequencing.
+
+### Abstraction Interface
+
+The following will be the new `SecretProvider` abstraction interface used by all Edgex services
+
+```go
+type SecretProvider interface {
+    // Stores new secrets into the service's exclusive SecretStore at the speficied path.
+	StoreSecrets(path string, secrets map[string]string) error
+	// Retrieves secrets from the service's exclusive SecretStore at the speficied path.
+	GetSecrets(path string, _ ...string) (map[string]string, error)
+    // Sets the secrets lastupdated time to current time. 
+    // Needed for InsecureSecrets support
+	SecretsUpdated()
+    // Returns the secrets last updated time
+	SecretsLastUpdated() time.Time
+}
+```
+
+> *Note: The `GetDatabaseCredentials` and `GetCertificateKeyPair` APIs have been removed. These are no longer needed since insecure database credentials will no longer be stored in the `DatabaseInfo` configuration and certificate key pairs are secrets like any others. This allows these secrets to be retrieved via the `GetSecrets` API.*
+
+### Implementation
+
+#### Factory Method and Bootstrap Handler
+
+The factory method and bootstrap handler will follow that currently in the Bootstrap implementation with some tweaks. Rather than putting the two split interfaces into the DIC, it will put just the single interface instance into the DIC. See details in the [Interfaces and factory methods](#interfaces-and-factory-methods) section above under **Existing Implementations**.
+
+#### Caching of Secrets
+
+Secrets will be cached as they are currently in the Application Service implementation
```

Comment:
How does the caching work? Is it the responsibility of the service to update the cache, or will it be handled automatically?

Reply:
Handled Automatically by the SecretProvider.